python/opae.admin/opae/admin/tools/fpgasupdate.py

```diff
@@ -351,27 +381,183 @@ def canonicalize_bdf(bdf):
+        if retries > max_retries:
+            return errno.ETIMEDOUT, 'Secure update timed out'
+        _, progress_code, _, _ = ioctl_status(sec_dev)
+
+    # IOCTL Call to kick off the write
+    fcntl.ioctl(sec_dev, IOCTL_FPGA_IMAGE_LOAD_WRITE, write_header, True)
+
+    LOG.info('preparing image file')
+    retries = 0
+    max_retries = 60 * 5
+    _, progress_code, _, _ = ioctl_status(sec_dev)
+    while progress_code in [FPGA_IMAGE_PROG_STARTING, FPGA_IMAGE_PROG_PREPARING]:
+        time.sleep(timeout)
+        retries += 1
+        if retries > max_retries:
+            return errno.ETIMEDOUT, 'Secure update timed out'
+        _, progress_code, _, _ = ioctl_status(sec_dev)
+
+    _, progress_code, _, err_code = ioctl_status(sec_dev)
+    if progress_code == FPGA_IMAGE_PROG_IDLE and err_code != FPGA_IMAGE_ERR_NONE:
+        return 1, fpga_err_to_string(err_code)
+
+    progress_cfg = {}
+    level = min([handler.level for handler in LOG.handlers])
+    if level < logging.INFO:
+        progress_cfg['log'] = LOG.debug
+    else:
+        progress_cfg['stream'] = sys.stdout
+
+    LOG.info('writing image file')
+    retries = 0
+    max_retries = 60 * 60 * 2
+    with progress(bytes=payload_size, **progress_cfg) as prg:
+        remaining_size, _, _, _ = ioctl_status(sec_dev)
+        while int(remaining_size) > 0:
+            time.sleep(timeout)
+            retries += 1
+            if retries > max_retries:
+                return errno.ETIMEDOUT, 'Secure update timed out'
+            prg.update(payload_size - int(remaining_size))
+            remaining_size, _, _, _ = ioctl_status(sec_dev)
+
+    _, progress_code, _, err_code = ioctl_status(sec_dev)
+    if progress_code == FPGA_IMAGE_PROG_IDLE and err_code != FPGA_IMAGE_ERR_NONE:
+        return 1, fpga_err_to_string(err_code)
+
+    if pac.fme.have_node('tcm'):
+        estimated_time = payload_size / DRAM_COPY_BPS + DRAM_COPY_OFFSET
+    else:
+        estimated_time = payload_size / FLASH_COPY_BPS
+    # over-estimate by 1.5 to account for flash performance degradation
+    estimated_time *= 1.5
+
+    LOG.info('programming image file')
+    interrupt_flag = False
+    retries = 0
+    max_retries = 60 * 60 * 3
+    with progress(time=estimated_time, **progress_cfg) as prg:
+        while progress_code in [FPGA_IMAGE_PROG_WRITING, FPGA_IMAGE_PROG_PROGRAMMING]:
+            try:
+                time.sleep(timeout)
+                retries += 1
+                if retries > max_retries:
+                    return errno.ETIMEDOUT, 'Secure update timed out'
+                prg.tick()
+                _, progress_code, _, _ = ioctl_status(sec_dev)
+                if progress_code is FPGA_IMAGE_PROG_PROGRAMMING and interrupt_flag:
+                    LOG.warning('Ignoring Ctrl+C: programming phase is not interruptable')
+                    interrupt_flag = False
+            except KeyboardInterrupt:
+                fcntl.ioctl(sec_dev, IOCTL_FPGA_IMAGE_LOAD_CANCEL)
```

Comment:
Need a try: here to guard against ioctl exceptions. If the ioctl() fails, an OSError exception is raised.